src/java.base/share/classes/com/sun/crypto/provider/ML_KEM.java

```diff
@@ -0,0 +1,632 @@
+                out.writeBytes(ps.get(i).Encode(d));
+            }
+            return out.toByteArray();
+        }
+
+        @Override
+        public String toString() {
+            var sb = new StringBuilder("Vec(");
+            for (var p : ps) {
+                sb.append(p).append(", ");
+            }
+            sb.append(")");
+            return sb.toString();
+        }
+
+    }
+
+    static byte[] EncodeVec(Vec v, int w) {
+        var out = new ByteArrayOutputStream();
+        for (var p : v.ps) {
+            out.writeBytes(Encode(p.cs, w));
+        }
+        return out.toByteArray();
+    }
+
+    static Vec DecodeVec(byte[] bs, int k, int w) {
+        var cs = Decode(bs, w);
+        var list = new ArrayList<Poly>();
+        for (var i = 0; i < k; i++) {
+            list.add(new Poly(Arrays.copyOfRange(cs, n * i, n * (i + 1))));
+        }
+        return new Vec(list);
+    }
+
+    static Poly DecodePoly(byte[] bs, int w) {
+        return new Poly(Decode(bs, w));
+    }
+
+    static class Matrix {
+        final List<Vec> cs;
+
+        Matrix(List<Vec> cs) {
+            this.cs = cs;
+        }
+
+        Vec MulNTT(Vec vec) {
+            return new Vec(cs.stream().map(r -> r.DotNTT(vec)).toList());
+        }
+
+        Matrix T() {
+            var k = cs.size();
+            var newCs = new ArrayList<Vec>();
+            for (var i = 0; i < k; i++) {
+                var newPs = new ArrayList<Poly>();
+                for (var j = 0; j < k; j++) {
+                    newPs.add(cs.get(j).ps.get(i));
+                }
+                newCs.add(new Vec(newPs));
+            }
+            return new Matrix(newCs);
+        }
+
+        @Override
+        public String toString() {
+            var sb = new StringBuilder("Matrix(");
+            for (var p : cs) {
+                sb.append(p).append(", ");
+            }
+            sb.append(")");
+            return sb.toString();
+        }
+    }
+
+    static Matrix sampleMatrix(byte[] rho, int k) {
+        var newCs = new ArrayList<Vec>();
+        for (var i = 0; i < k; i++) {
+            var newPs = new ArrayList<Poly>();
+            for (var j = 0; j < k; j++) {
+                newPs.add(sampleUniform(xof(rho, j, i)));
+            }
+            newCs.add(new Vec(newPs));
+        }
+        return new Matrix(newCs);
+    }
+
+
+    static Vec sampleNoise(byte[] sigma, int eta, int offset, int k) {
+        return new Vec(IntStream.range(0, k).mapToObj(i -> CBD(PRF1(sigma, i + offset, 64 * eta), eta)).toList());
+    }
+
+
+    static byte[] constantTimeSelectOnEquality(byte[] a, byte[] b, byte[] ifEq, byte[] ifNeq) {
+        //    # WARNING! In production code this must be done in a
+        //    # data-independent constant-time manner, which this implementation
+        //    # is not. In fact, many more lines of code in this
+        //    # file are not constant-time.
+        return MessageDigest.isEqual(a, b) ? ifEq : ifNeq;
```

Comment:
I was under the impression that `MessageDigest.isEqual()` runs in constant time for any value of `a` at a given length independent of `b`'s length or value.  Is this an old comment?  Or is the concern about `b` itself?

Reply:
This is not Ferenc's version, but a very straightforward translation of the spec. Actually the comparison is constant time here. Please do not review it unless you just want to learn about the algorithm.